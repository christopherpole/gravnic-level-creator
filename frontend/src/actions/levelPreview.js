import { convertTilesToGameState } from 'gravnic-game';

export const CHANGE_GRAVITY_DIRECTION = 'CHANGE_GRAVITY_DIRECTION';
export const UPDATE_GAME_STATE = 'UPDATE_GAME_STATE';
export const ENTITIES_STOPPED_MOVING = 'ENTITIES_STOPPED_MOVING';
export const PREVIEW_LEVEL = 'PREVIEW_LEVEL';
export const EDIT_LEVEL = 'EDIT_LEVEL';

export const previewLevel = () => (dispatch, getState) => {
  const { tiles } = getState().levelEditor;
  const gameState = convertTilesToGameState(tiles);

  dispatch({
    type: PREVIEW_LEVEL,
    gameState,
  });
};

export const editLevel = () => ({
  type: EDIT_LEVEL,
});

export const changeGravityDirection = direction => ({
  type: CHANGE_GRAVITY_DIRECTION,
  direction,
});

export const updateGameState = gameState => ({
  type: UPDATE_GAME_STATE,
  gameState,
});

export const entitiesStoppedMoving = () => ({
  type: ENTITIES_STOPPED_MOVING,
});
