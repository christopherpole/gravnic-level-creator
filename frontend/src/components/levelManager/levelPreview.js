import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Tile from '../common/tile';

export const Wrapper = styled.div`
  position: relative;
  height: 0;
  padding-bottom: 100%;
`;

export const WrapperInner = styled.div`
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  grid-template-rows: repeat(10, 1fr);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`;

export const TileContainer = styled.div`
  position: relative;
`;

export const LevelPreview = ({ tiles }) => (
  <Wrapper>
    <WrapperInner>
      {tiles.map(tile => (
        <TileContainer key={tile.position}>
          <Tile id={tile.selectedTileId} />
        </TileContainer>
      ))}
    </WrapperInner>
  </Wrapper>
);

LevelPreview.propTypes = {
  tiles: PropTypes.array.isRequired,
};

export default LevelPreview;
