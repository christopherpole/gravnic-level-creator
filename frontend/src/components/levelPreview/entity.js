import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Tile from '../common/tile';
import { GRID_SIZE } from '../../config/settings';

export const Wrapper = styled.div`
  height: ${props => props.heightPercentage}%;
  width: ${props => props.widthPercentage}%;
  position: absolute;
  left: ${props => props.xPos}%;
  top: ${props => props.yPos}%;
`;

export const Entity = ({ entityId, xPos, yPos }) => (
  <Wrapper
    xPos={xPos}
    yPos={yPos}
    heightPercentage={100 / GRID_SIZE}
    widthPercentage={100 / GRID_SIZE}
  >
    <Tile tileId={entityId} />
  </Wrapper>
);

Entity.propTypes = {
  entityId: PropTypes.number.isRequired,
  xPos: PropTypes.number.isRequired,
  yPos: PropTypes.number.isRequired,
};

export default Entity;
