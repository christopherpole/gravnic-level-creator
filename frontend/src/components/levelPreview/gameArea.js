import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Entity from './entity';
import { GRID_SIZE } from '../../config/settings';

export const Wrapper = styled.div``;

export const GameArea = ({ gameState }) => (
  <Wrapper>
    {gameState.map((gameStateRow, i) =>
      gameStateRow.map((entityId, j) => (
        <Entity
          xPos={j * (100 / GRID_SIZE)}
          yPos={i * (100 / GRID_SIZE)}
          key={i + j}
          entityId={entityId}
        />
      )),
    )}
  </Wrapper>
);

GameArea.propTypes = {
  gameState: PropTypes.array.isRequired,
};

const mapStateToProps = state => ({
  gameState: state.levelPreview.gameState,
});

export default connect(mapStateToProps)(GameArea);
