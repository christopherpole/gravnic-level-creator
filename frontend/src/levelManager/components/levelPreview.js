import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Tile from 'common/tile';

export const Wrapper = styled.div`
  position: relative;
  height: 0;
  padding-bottom: 100%;
`;

export const WrapperInner = styled.div`
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  grid-template-rows: repeat(10, 1fr);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`;

export const TileContainer = styled.div`
  position: relative;
`;

export const LevelPreview = ({ tiles }) => (
  <Wrapper className="level-preview">
    <WrapperInner>
      {tiles.map(tile => (
        <TileContainer className="preview-tile" key={tile.position}>
          <Tile tileId={tile.selectedTileId} />
        </TileContainer>
      ))}
    </WrapperInner>
  </Wrapper>
);

LevelPreview.propTypes = {
  tiles: PropTypes.array.isRequired,
};

export default LevelPreview;
