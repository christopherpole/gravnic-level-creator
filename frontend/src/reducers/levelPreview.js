import {
  PREVIEW_LEVEL,
  EDIT_LEVEL,
  CHANGE_GRAVITY_DIRECTION,
  UPDATE_GAME_STATE,
  ENTITIES_STOPPED_MOVING,
} from '../actions/levelPreview';

export const initialState = {
  previewing: false,
  gameState: null,
  gravityDirection: null,
  entitiesMoving: false,
  gameHistory: [],
};

export default function levelPreviewReducer(state = initialState, action) {
  switch (action.type) {
    case PREVIEW_LEVEL: {
      return {
        ...state,
        previewing: true,
        gameState: action.gameState,
      };
    }

    case EDIT_LEVEL: {
      return {
        ...state,
        previewing: false,
      };
    }

    case CHANGE_GRAVITY_DIRECTION: {
      const newGameHistory = [...state.gameHistory];
      newGameHistory.push([]);

      return {
        ...state,
        gravityDirection: action.direction,
        entitiesMoving: true,
        gameHistory: newGameHistory,
      };
    }

    case UPDATE_GAME_STATE: {
      const newGameHistory = [...state.gameHistory];
      newGameHistory[newGameHistory.length - 1].push(action.gameState);

      return {
        ...state,
        gameState: action.gameState,
        gameHistory: newGameHistory,
      };
    }

    case ENTITIES_STOPPED_MOVING: {
      return {
        ...state,
        entitiesMoving: false,
      };
    }

    default: {
      return state;
    }
  }
}
