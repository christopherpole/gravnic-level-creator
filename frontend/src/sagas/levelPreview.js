import { calulateNextGameState } from 'gravnic-game';
import { takeLatest, delay } from 'redux-saga';
import { call, put, select } from 'redux-saga/effects';

import {
  CHANGE_GRAVITY_DIRECTION,
  updateGameState,
  entitiesStoppedMoving,
} from '../actions/levelPreview';
import { ENTITY_MOVE_SPEED } from '../config/settings';

export function* changeGravityDirectionSaga(action) {
  const state = yield select();
  let { gameState } = state.levelPreview;

  do {
    gameState = calulateNextGameState(gameState, action.direction);

    if (gameState) {
      yield put(updateGameState(gameState));
      yield call(delay, ENTITY_MOVE_SPEED);
    }
  } while (gameState);

  yield put(entitiesStoppedMoving());
}

export default function* levelManagerSagas() {
  yield takeLatest(CHANGE_GRAVITY_DIRECTION, changeGravityDirectionSaga);
}
