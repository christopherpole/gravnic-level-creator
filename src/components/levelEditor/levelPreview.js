import React from 'react';
import styled from 'styled-components';

export const Wrapper = styled.div`
  height: 0;
  padding-bottom: 100%;
  position: relative;
`;

export const WrapperInner = styled.div`
  background: white;
  color: black;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  font-weight: bold;
`;

export const LevelPreview = () => (
  <Wrapper id="level-preview">
    <WrapperInner>
      <p>This is a preview</p>
    </WrapperInner>
  </Wrapper>
);

export default LevelPreview;
