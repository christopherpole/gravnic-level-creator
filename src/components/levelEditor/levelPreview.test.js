import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { LevelPreview } from './levelPreview';

configure({ adapter: new Adapter() });

describe('The tile selector', () => {
  let props;

  beforeEach(() => {
    props = {
    };
  });

  it('Renders without exploding', () => {
    const levelPrevieww = shallow(<LevelPreview {...props} />);

    expect(levelPrevieww).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const levelPrevieww = shallow(<LevelPreview {...props} />);

    expect(toJson(levelPrevieww)).toMatchSnapshot();
  });
});
