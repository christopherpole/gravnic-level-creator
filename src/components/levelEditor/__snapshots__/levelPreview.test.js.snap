// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tile selector Matches the current snapshot 1`] = `
<styled.div
  id="level-preview"
>
  <styled.div>
    <p>
      This is a preview
    </p>
  </styled.div>
</styled.div>
`;
